Add support for Add Sve.GatherVectorUInt*ZeroExtendFirstFaulting()

SwapnilGaikwad opened the pull request:
Contribute towards https://github.com/dotnet/runtime/issues/99957

@a74nh @kunalspathak @dotnet/arm64-contrib @arch-arm64-sve

TIHan committed:
Initial work

TIHan committed:
Merge remote-tracking branch 'upstream/main' into sve-ffr-part1

TIHan committed:
FirstFaulting partially works

TIHan committed:
Added template

TIHan committed:
Trying to test first-faulting behavior

TIHan committed:
Using BoundedMemory to test FirstFaulting behavior for LoadVector.

TIHan committed:
Fix size in validation

TIHan committed:
Added more helper functions. Added conditional select tests for LoadVectorFirstFaulting.

TIHan committed:
Added first-faulting behavior tests for GatherVectorFirstFaulting

TIHan committed:
Merging with main

TIHan committed:
Added GetFfr suffix-style APIs

TIHan committed:
Fixing GatherVector tests

TIHan committed:
Formatting

TIHan committed:
Feedback

TIHan committed:
Merge remote-tracking branch 'upstream/main' into sve-ffr-part1

TIHan committed:
Feedback

tannergooding committed:
Ensure the P/Invokes are blittable

TIHan committed:
Merging

TIHan committed:
Merge remote-tracking branch 'upstream/main' into sve-ffr-part1

TIHan committed:
Fix build

TIHan committed:
Remove checking for zeroes after the fault

SwapnilGaikwad committed:
Add Sve.GatherVectorUInt*ZeroExtendFirstFaulting()

TIHan committed:
Added GatherVectorFirstFaultingVectorBases test template, but currently without the FirstFaulting test. Added SveFfrTest template.

TIHan committed:
Mark GetFfr methods as side-effectful

SwapnilGaikwad commented:
Occasionally, it fails on a stress test. More likely to do with the test than the functionality. Looking into it.
 
<details>
<summary> Stress test results </summary>

```
===================Running default===================
------------------- {} -------------------
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtend_Indices_int_int() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtend_Indices_uint_int() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtend_Indices_int_uint() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtend_Indices_uint_uint() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtend_Indices_long_long() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtend_Indices_ulong_long() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtend_Indices_long_ulong() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtend_Indices_ulong_ulong() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtend_Bases_long_ulong() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtend_Bases_ulong_ulong() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtend_Indices_int_int() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtend_Indices_uint_int() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtend_Indices_int_uint() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtend_Indices_uint_uint() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtend_Indices_long_long() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtend_Indices_ulong_long() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtend_Indices_long_ulong() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtend_Indices_ulong_ulong() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtend_Indices_long_long() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtend_Indices_int_int() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtend_Indices_ulong_long() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtend_Indices_uint_int() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtend_Indices_long_ulong() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtend_Indices_int_uint() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtend_Indices_ulong_ulong() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtend_Indices_uint_uint() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtend_Bases_long_ulong() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtend_Bases_ulong_ulong() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtend_Indices_long_long() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtend_Indices_int_int() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtend_Indices_ulong_long() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtend_Indices_uint_int() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtend_Indices_long_ulong() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtend_Indices_int_uint() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtend_Indices_ulong_ulong() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtend_Indices_uint_uint() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Indices_int_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Indices_int_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Indices_long_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Indices_long_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Indices_uint_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Indices_uint_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Indices_ulong_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Indices_ulong_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Indices_int_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Indices_int_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Indices_long_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Indices_long_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Indices_uint_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Indices_uint_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Indices_ulong_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Indices_ulong_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Bases_long_ulong() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Bases_ulong_ulong() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Bases_long_ulong() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Bases_ulong_ulong() : 15
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtendFirstFaulting_offsets_int_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtendFirstFaulting_offsets_int_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtendFirstFaulting_offsets_long_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtendFirstFaulting_offsets_long_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtendFirstFaulting_offsets_uint_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtendFirstFaulting_offsets_uint_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtendFirstFaulting_offsets_ulong_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtendFirstFaulting_offsets_ulong_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtendFirstFaulting_offsets_int_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtendFirstFaulting_offsets_int_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtendFirstFaulting_offsets_long_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtendFirstFaulting_offsets_long_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtendFirstFaulting_offsets_uint_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtendFirstFaulting_offsets_uint_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtendFirstFaulting_offsets_ulong_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtendFirstFaulting_offsets_ulong_ulong() : 13
===================Running jitstress===================
------------------- {'JitMinOpts': '1'} -------------------
------------------- {'JitStress': '1'} -------------------
Test failed:
..........................................
..........................................
Sve.GatherVectorUInt32WithByteOffsetsZeroExtendFirstFaulting<UInt32>(UInt32, UInt32, UInt32): RunBasicScenario_LoadFirstFaulting failed:
       firstOp: (0, 1, 1, 1)
      secondOp: (507388, 260062, 196593, 0, 3201305514, 3490442496, 81, 0, 1302197768, 64245, 5, 0, 219521864, 64245, 219521720, 64245, 219519480, 64245, 219519696, 64245, 219519768, 64245, 219519840, 64245, 219519768, 64245, 33, 0, 160739304, 64245, 0, 0, 32, 0, 49, 0, 3534340144, 65535, 205629744, 49748, 23, 2987803337, 4, 17, 17, 0, 49, 0, 3534343296, 65535, 205055984, 49748, 23, 2987803337, 4, 17, 17, 0, 49, 0, 3534342464, 65535, 205785256, 49748, 23, 2987803337, 4, 17, 17, 0, 49, 0, 221709184, 49748, 219217920, 64245, 65536, 0, 61214720, 0, 0, 0, 145, 0, 0, 49752, 0, 0, 5, 0, 33555895, 16777217, 67113628, 167772166, 0, 0, 67113630, 167772170, 67113627, 167772164, 0, 1879048197, 0, 0, 5, 1879048199, 1879232422, 1879048193, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1879048201, 49, 0, 3534342832, 65535, 206275800, 49748, 23, 2987803337, 4, 17, 17, 0, 49, 0, 3534343104, 65535, 203496456, 49748, 23, 2987803337, 4, 17, 17, 3490442496, 49, 0, 219964400, 49748, 20, 0, 221880000, 49748, 1, 0, 3201305514, 3490442496, 49, 0, 221599808, 49748, 69, 0, 221859152, 49748, 2, 0, 0, 0, 49, 0, 3534342960, 65535, 206252216, 49748, 23, 2987803337, 4, 17, 17, 0, 49, 0, 221591904, 49748, 19, 0, 221811312, 49748, 1, 0, 0, 0, 161, 0, 1302197768, 64245, 16, 0, 219483640, 64245, 205457336, 64245, 219483208, 64245, 219483280, 64245, 219483352, 64245, 219481040, 64245, 219481112, 64245, 219481256, 64245, 219481328, 64245, 219484416, 64245, 219481760, 64245, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 10, 0, 65, 0, 1302197768, 64245, 3, 0, 218946032, 64245, 218919336, 64245, 219516112, 64245, 219757664, 64245, 219509904, 64245, 161, 0, 1302197768, 64245, 16, 0, 161131160, 64245, 163097208, 64245, 163694496, 64245, 215025752, 64245, 215798824, 64245, 215949864, 64245, 215963952, 64245, 218421152, 64245, 218425848, 64245, 218642784, 64245, 218647480, 64245, 219488960, 64245, 219495776, 64245, 219502192, 64245, 219514232, 64245, 219780064, 64245, 0, 0, 97, 0, 1302197768, 64245, 5, 64245, 219796392, 64245, 219796176, 64245, 219793936, 64245, 219794224, 64245, 219794296, 64245, 219794368, 64245, 0, 0, 2147500033, 0, 96, 0, 49, 0, 3534343280, 65535, 206283080, 49748, 23, 2987803337, 4, 17, 17, 64245, 49, 0, 3534342992, 65535, 205155528, 49748, 23, 2987803337, 4, 17, 17, 64245, 49, 0, 1302197768, 64245, 1, 0, 219794800, 64245, 219794728, 64245, 2, 0, 49, 0, 3534343424, 65535, 204530624, 49748, 23, 2987803337, 4, 17, 17, 0, 49, 0, 3534340320, 65535, 203049928, 49748, 23, 2987803337, 4, 17, 17, 0, 49, 0, 3534343264, 65535, 205205872, 49748, 23, 2987803337, 4, 17, 17, 0, 49, 0, 3534343296, 65535, 206573848, 49748, 23, 2987803337, 4, 17, 17, 0, 49, 0, 3534343088, 65535, 205159944, 49748, 23, 2987803337, 4, 17, 17, 0, 545, 0, 1302201392, 64245, 1, 0, 167928080, 64245, 167928080, 64245, 1, 30, 221642752, 49748, 0, 0, 0, 0, 167927728, 64245, 167927728, 64245, 167926768, 64245, 167926768, 64245, 167926896, 64245, 167926896, 64245, 167926192, 64245, 167926192, 64245, 167926256, 64245, 167926256, 64245, 167926320, 64245, 167926320, 64245, 167926384, 64245, 167926384, 64245, 167926448, 64245, 167926448, 64245, 167926512, 64245, 167926512, 64245, 167926576, 64245, 167926576, 64245, 167926640, 64245, 167926640, 64245, 167926704, 64245, 167926704, 64245, 167927024, 64245, 167927024, 64245, 167927088, 64245, 167927088, 64245, 167927152, 64245, 167927152, 64245, 167927216, 64245, 167927216, 64245, 167927280, 64245, 167927280, 64245, 167927344, 64245, 167927344, 64245, 167927408, 64245, 167927408, 64245, 167927472, 64245, 167927472, 64245, 167927536, 64245, 167927536, 64245, 167927600, 64245, 167927600, 64245, 167927664, 64245, 167927664, 64245, 167927856, 64245, 167927856, 64245, 167927920, 64245, 167927920, 64245, 167927984, 64245, 167927984, 64245, 167926832, 64245, 167926832, 64245, 167926960, 64245, 167926960, 64245, 167927792, 64245, 167927792, 64245, 0, 0, 49, 0, 3534343392, 65535, 204488472, 49748, 23, 2987803337, 4, 17, 17, 0, 49, 0, 3534343264, 65535, 206281624, 49748, 23, 2987803337, 4, 17, 17, 0, 33, 0, 221850080, 49748, 221848128, 49748, 32, 0, 48, 0, 3534343024, 65535, 205033344, 49748, 23, 2987803337, 4, 17, 17, 0, 49, 0, 3534343280, 65535, 203090712, 49748, 23, 2987803337, 4, 17, 17, 64245, 49, 0, 1302197768, 64245, 1, 64245, 219789840, 64245, 205458840, 64245, 3532707329, 4071196225, 49, 0, 1302197768, 64245, 1, 64245, 219789048, 64245, 205458840, 64245, 3594453120, 4181724065, 49, 0, 3534343424, 65535, 203014224, 49748, 23, 2987803337, 4, 17, 17, 4181721120, 49, 0, 3534343088, 65535, 203509944, 49748, 23, 2987803337, 4, 17, 17, 0, 97, 0, 678375017, 49752, 2037907264, 2066541974, 0, 0, 221682544, 49748, 0, 64245, 221682544, 49748, 0, 0, 0, 64245, 0, 0, 2147500033, 0, 96, 0, 97, 0, 1302197768, 64245, 8, 0, 213449680, 64245, 218913600, 64245, 219497656, 64245, 219516112, 64245, 219521720, 64245, 219757664, 64245, 219781944, 64245, 219791432, 64245, 219796176, 64245, 49, 0, 3534343280, 65535, 204179208, 49748, 23, 2987803337, 4, 17, 17, 4181722806, 49, 0, 3534340208, 65535, 206724400, 49748, 23, 2987803337, 4, 17, 17, 2432746400, 49, 0, 220601216, 49748, 5, 0, 220603616, 49748, 0, 0, 3201305514, 3490442496, 49, 0, 3534343280, 65535, 206566912, 49748, 23, 2987803337, 4, 17, 17, 4070654145, 49, 0, 221860224, 49748, 19, 0, 221615280, 49748, 1, 0, 0, 0, 49, 0, 3534342864, 65535, 205917528, 49748, 23, 2987803337, 4, 17, 17, 0, 401, 0, 625005417, 12, 2037907264, 2066541974, 2835635189, 2835702775, 2835770361, 2835837947, 2835905533, 2432828413, 3532312327, 4070721159, 4074495495, 4076033927, 4162487207, 1016660896, 2852127731, 2852193271, 2852258805, 704840692, 2852389880, 2852455417, 2852520950, 3533213978, 4070656922, 4074725050, 2853831648, 3531604097, 2536214956, 2852127739, 320831105, 301990945, 185860129, 318929953, 2470476833, 2853831648, 2536214948, 2852127732, 2853831648, 3531604097, 2536214944, 2852127740, 2853831648, 3531604097, 2536214940, 2852127738, 1035993840, 1031800688, 3107981952, 1895825439, 1409287369, 2432713344, 3107981954, 1394504770, 704775138, 2853503969, 2530021462, 1035993856, 1031800704, 1035993888, 1031800640, 2853372896, 2853897185, 2853438434, 2853962723, 2853831652, 1020855200, 2853569509, 3533587718, 4070679174, 4074725030, 4181721286, 3532312336, 4070721168, 4074495504, 4076033936, 4166681521, 3943760415, 1409286208, 2536214881, 2840099837, 2840032251, 2839964665, 2839897079, 2839829493, 2839761907, 2432844799, 3592356032, 2536214908, 3558866944, 407896152, 8388686, 3873837282, 3370575590, 3840149766, 0, 3840149766, 0, 49, 0, 3534343280, 65535, 204840592, 49748, 23, 2987803337, 4, 17, 17, 0, 49, 0, 1302197768, 64245, 1, 0, 219790056, 64245, 219789984, 64245, 0, 0, 49, 0, 3534343152, 65535, 205744976, 49748, 23, 2987803337, 4, 17, 17, 0, 49, 0, 3534343184, 65535, 206602712, 49748, 23, 2987803337, 4, 17, 17, 0, 49, 0, 3534343408, 65535, 203671184, 49748, 23, 2987803337, 4, 17, 17, 0, 49, 0, 3534343248, 65535, 203700480, 49748, 23, 2987803337, 4, 17, 17, 0, 49, 0, 1302197768, 64245, 1, 64245, 219803280, 64245, 205459056, 64245, 19, 0, 49, 0, 3534343424, 65535, 203726864, 49748, 23, 2987803337, 4, 17, 17, 64245, 49, 0, 3534339984, 65535, 206807280, 49748, 23, 2987803337, 4, 17, 17, 0, 97, 0, 221612448, 49748, 221848736, 49748, 221848752, 49748, 221683856, 49748, 1, 0, 221683856, 49748, 221848760, 49748, 3, 64245, 221848752, 49748, 0, 1, 96, 0, 48, 0, 678391450, 49752, 2037907264, 2066541974, 219805736, 64245, 205459056, 64245, 219805664, 64245, 49, 0, 3534343200, 65535, 203986400, 49748, 23, 2987803337, 4, 17, 17, 0, 49, 0)
       thirdOp: (251, 0, 250, 853)
        result: (0, 507388, 16, 80)
   faultResult: (<16843009, 16843009, 16843009, 16843009>)
..........................................
System.Exception: One or more scenarios did not complete as expected.
   at JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtendFirstFaulting_offsets_uint_uint() in /home/swagai01/dotnet/runtime/artifacts/tests/coreclr/obj/linux.arm64.Checked/Managed/JIT/HardwareIntrinsics/Arm/Sve/Sve_ro/Sve_ro/gen/Sve.GatherVectorUInt32WithByteOffsetsZeroExtendFirstFaulting.offsets.uint.uint.cs:line 87
   at Program.<<Main>$>g__TestExecutor3229|0_3230(StreamWriter tempLogSw, StreamWriter statsCsvSw, <>c__DisplayClass0_0&) in /home/swagai01/dotnet/runtime/artifacts/tests/coreclr/obj/linux.arm64.Checked/Managed/JIT/HardwareIntrinsics/HardwareIntrinsics_Arm_ro/generated/XUnitWrapperGenerator/XUnitWrapperGenerator.XUnitWrapperGenerator/FullRunner.g.cs:line 81553
------------------- {'JitStress': '2'} -------------------
------------------- {'JitStress': '1', 'TieredCompilation': '1'} -------------------
------------------- {'JitStress': '2', 'TieredCompilation': '1'} -------------------
------------------- {'TailcallStress': '1'} -------------------
------------------- {'ReadyToRun': '0'} -------------------
===================Running jitstressregs===================
------------------- {'JitStressRegs': '1'} -------------------
------------------- {'JitStressRegs': '2'} -------------------
------------------- {'JitStressRegs': '3'} -------------------
------------------- {'JitStressRegs': '4'} -------------------
------------------- {'JitStressRegs': '8'} -------------------
------------------- {'JitStressRegs': '0x10'} -------------------
------------------- {'JitStressRegs': '0x80'} -------------------
------------------- {'JitStressRegs': '0x1000'} -------------------
------------------- {'JitStressRegs': '0x2000'} -------------------
===================Running jitstress2-jitstressregs===================
------------------- {'JitStress': '2', 'JitStressRegs': '1'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '2'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '3'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '4'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '8'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x10'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x80'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x1000'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x2000'} -------------------
```
</details>

kunalspathak commented:
i have seen some failures on other `GatherVector` that Alan added, so probably testing issue.

kunalspathak commented on src/coreclr/jit/lsraarm64.cpp:
```diff
@@ -1992,9 +1992,13 @@ int LinearScan::BuildHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree, int* pDstCou
                 case NI_Sve_GatherVectorInt32WithByteOffsetsSignExtend:
                 case NI_Sve_GatherVectorSByteSignExtend:
                 case NI_Sve_GatherVectorUInt16WithByteOffsetsZeroExtend:
+                case NI_Sve_GatherVectorUInt16WithByteOffsetsZeroExtendFirstFaulting:
```
@TIHan - did you miss adding `GatherVectorFirstFaulting` and `LoadVectorFirstFaulting`? Can you check how is RefPosition for `op2` is created? I guess we are missing for bunch of others too.

kunalspathak replied:
@mikabl-arm - same for you in https://github.com/dotnet/runtime/pull/104964

kunalspathak commented on src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs:
```diff
@@ -1017,7 +1017,7 @@
     ("LoadVectorx4Test.template",         new Dictionary<string, string> { ["TestName"] = "Load4xVector64AndUnzipShort",                                                                           ["Isa"] = "AdvSimd", ["LoadIsa"] = "AdvSimd",       ["Method"] = "Load4xVector64AndUnzip",                                                                                                                   ...
     ("LoadVectorx4Test.template",         new Dictionary<string, string> { ["TestName"] = "Load4xVector64AndUnzipUInt32",                                                                          ["Isa"] = "AdvSimd", ["LoadIsa"] = "AdvSimd",       ["Method"] = "Load4xVector64AndUnzip",                                                                                                                   ...
     ("LoadVectorx4Test.template",         new Dictionary<string, string> { ["TestName"] = "Load4xVector64AndUnzipInt32",                                                                           ["Isa"] = "AdvSimd", ["LoadIsa"] = "AdvSimd",       ["Method"] = "Load4xVector64AndUnzip",                                                                                                                   ...
-    ("LoadVectorx4Test.template",         new Dictionary<string, string> { ["TestName"] = "Load4xVector64AndUnzipFloat",                                                                           ["Isa"] = "AdvSimd", ["LoadIsa"] = "AdvSimd",       ["Method"] = "Load4xVector64AndUnzip",                                                                                                                   ...
```
there are several space changes in this file. can you please revert them all?

SwapnilGaikwad replied:
Done

kunalspathak reviewed (commented):
overall looks good.

SwapnilGaikwad committed:
Add firstFaultingLoad test for VectorBases

SwapnilGaikwad committed:
Remove redundant spaces

kunalspathak committed:
Add tracking of FFR register

somewhat workable

code cleanup

Remove FFR

Add all the GetFfr*

wip

Work with MskCns() model

Use physReg approach

Remove commented prototypes

working

Remove bunch of unnecessary code

Remove SpecialImport from GetFFR/SetFFR/LoadFirstFaulting

some more code cleanup

some fixup

SwapnilGaikwad committed:
Add remaining Sve.GatherVectorFirstFaulting()

TIHan committed:
Verifying expected fault result. Test weaks.

TIHan committed:
Merging with main

TIHan committed:
Fix build

SwapnilGaikwad committed:
Fix intrinsic order

SwapnilGaikwad committed:
Merge will-ffr changes

SwapnilGaikwad committed:
Fix GetFFR type

SwapnilGaikwad committed:
Fix formatting issues

SwapnilGaikwad committed:
Fix PoisonPage configuration while creating BoundedMemory

SwapnilGaikwad committed:
Fix PoisonPage configuration while creating BoundedMemory

SwapnilGaikwad committed:
Use mmap() instead of memalign() for memory allocation

SwapnilGaikwad committed:
Use mmap() instead of memalign() for memory allocation

kunalspathak committed:
Change condition for PhysReg

kunalspathak committed:
jit format

kunalspathak committed:
review feedback

kunalspathak committed:
unspill for LoadVectorFirstFaulting as well

TIHan committed:
Merging with Kunal's FFR changes

SwapnilGaikwad committed:
Show error codes on failing failure

SwapnilGaikwad committed:
Show error codes on failing failure

TIHan committed:
Merging with main

TIHan committed:
Feedback

TIHan committed:
Feedback

TIHan committed:
Feedback

TIHan committed:
Feedback

SwapnilGaikwad committed:
Merge Will's FFR changes

kunalspathak committed:
Handle FFR correctly

kunalspathak committed:
reuse some of the code

kunalspathak committed:
Handle the special effect for SetFfr

kunalspathak committed:
some fixes + test coverage

SwapnilGaikwad committed:
Incorporate changes to preserve FFR correctly

kunalspathak committed:
do not zero init lvaFfrRegister

kunalspathak committed:
reverted local change

kunalspathak committed:
fix build break

SwapnilGaikwad committed:
Fix the arg count for GatherVector*firstFaulting in intrinsics table

SwapnilGaikwad commented:
The stress tests are passing now.

<details>
<summary>All Stress tests are passing</summary>

```
===================Running default===================
------------------- {} -------------------

Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Bases_double_ulong() : 16
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Bases_long_ulong() : 16
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Bases_ulong_ulong() : 16
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Indices_float_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Indices_int_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Indices_uint_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Indices_float_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Indices_int_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Indices_uint_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Indices_double_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Indices_long_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Indices_ulong_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Indices_double_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Indices_long_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Indices_ulong_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorByteZeroExtendFirstFaulting_Indices_int_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorByteZeroExtendFirstFaulting_Indices_int_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorByteZeroExtendFirstFaulting_Indices_long_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorByteZeroExtendFirstFaulting_Indices_long_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorByteZeroExtendFirstFaulting_Indices_uint_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorByteZeroExtendFirstFaulting_Indices_uint_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorByteZeroExtendFirstFaulting_Indices_ulong_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorByteZeroExtendFirstFaulting_Indices_ulong_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16SignExtendFirstFaulting_Indices_int_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16SignExtendFirstFaulting_Indices_int_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16SignExtendFirstFaulting_Indices_long_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16SignExtendFirstFaulting_Indices_long_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16SignExtendFirstFaulting_Indices_uint_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16SignExtendFirstFaulting_Indices_uint_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16SignExtendFirstFaulting_Indices_ulong_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16SignExtendFirstFaulting_Indices_ulong_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16WithByteOffsetsSignExtendFirstFaulting_offsets_int_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16WithByteOffsetsSignExtendFirstFaulting_offsets_int_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16WithByteOffsetsSignExtendFirstFaulting_offsets_long_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16WithByteOffsetsSignExtendFirstFaulting_offsets_long_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16WithByteOffsetsSignExtendFirstFaulting_offsets_uint_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16WithByteOffsetsSignExtendFirstFaulting_offsets_uint_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16WithByteOffsetsSignExtendFirstFaulting_offsets_ulong_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16WithByteOffsetsSignExtendFirstFaulting_offsets_ulong_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt32SignExtendFirstFaulting_Indices_long_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt32SignExtendFirstFaulting_Indices_long_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt32SignExtendFirstFaulting_Indices_ulong_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt32SignExtendFirstFaulting_Indices_ulong_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt32WithByteOffsetsSignExtendFirstFaulting_offsets_long_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt32WithByteOffsetsSignExtendFirstFaulting_offsets_long_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt32WithByteOffsetsSignExtendFirstFaulting_offsets_ulong_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt32WithByteOffsetsSignExtendFirstFaulting_offsets_ulong_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorSByteSignExtendFirstFaulting_Indices_int_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorSByteSignExtendFirstFaulting_Indices_int_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorSByteSignExtendFirstFaulting_Indices_long_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorSByteSignExtendFirstFaulting_Indices_long_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorSByteSignExtendFirstFaulting_Indices_uint_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorSByteSignExtendFirstFaulting_Indices_uint_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorSByteSignExtendFirstFaulting_Indices_ulong_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorSByteSignExtendFirstFaulting_Indices_ulong_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtendFirstFaulting_offsets_int_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtendFirstFaulting_offsets_int_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtendFirstFaulting_offsets_long_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtendFirstFaulting_offsets_long_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtendFirstFaulting_offsets_uint_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtendFirstFaulting_offsets_uint_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtendFirstFaulting_offsets_ulong_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtendFirstFaulting_offsets_ulong_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Indices_int_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Indices_int_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Indices_long_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Indices_long_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Indices_uint_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Indices_uint_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Indices_ulong_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Indices_ulong_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtendFirstFaulting_offsets_int_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtendFirstFaulting_offsets_int_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtendFirstFaulting_offsets_long_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtendFirstFaulting_offsets_long_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtendFirstFaulting_offsets_uint_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtendFirstFaulting_offsets_uint_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtendFirstFaulting_offsets_ulong_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtendFirstFaulting_offsets_ulong_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Indices_int_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Indices_int_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Indices_long_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Indices_long_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Indices_uint_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Indices_uint_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Indices_ulong_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Indices_ulong_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorByteZeroExtendFirstFaulting_Bases_long_ulong() : 16
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorByteZeroExtendFirstFaulting_Bases_ulong_ulong() : 16
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16SignExtendFirstFaulting_Bases_long_ulong() : 16
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16SignExtendFirstFaulting_Bases_ulong_ulong() : 16
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Bases_long_ulong() : 16
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Bases_ulong_ulong() : 16
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Bases_long_ulong() : 16
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Bases_ulong_ulong() : 16
===================Running jitstress===================
------------------- {'JitMinOpts': '1'} -------------------
------------------- {'JitStress': '1'} -------------------
------------------- {'JitStress': '2'} -------------------
------------------- {'JitStress': '1', 'TieredCompilation': '1'} -------------------
------------------- {'JitStress': '2', 'TieredCompilation': '1'} -------------------
------------------- {'TailcallStress': '1'} -------------------
------------------- {'ReadyToRun': '0'} -------------------
===================Running jitstressregs===================
------------------- {'JitStressRegs': '1'} -------------------
------------------- {'JitStressRegs': '2'} -------------------
------------------- {'JitStressRegs': '3'} -------------------
------------------- {'JitStressRegs': '4'} -------------------
------------------- {'JitStressRegs': '8'} -------------------
------------------- {'JitStressRegs': '0x10'} -------------------
------------------- {'JitStressRegs': '0x80'} -------------------
------------------- {'JitStressRegs': '0x1000'} -------------------
------------------- {'JitStressRegs': '0x2000'} -------------------
===================Running jitstress2-jitstressregs===================
------------------- {'JitStress': '2', 'JitStressRegs': '1'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '2'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '3'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '4'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '8'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x10'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x80'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x1000'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x2000'} -------------------
```
</details>

SwapnilGaikwad committed:
Merge main

SwapnilGaikwad committed:
Merge main

SwapnilGaikwad commented:
@TIHan @amanasifkhalid @a74nh 
As both #105595 and #104964 merged, I have rebased this PR and it's now ready for review.

<details>
<summary> Stress tests for the new GatherVector*FirstFaulting APIs are passing </summary>

```
===================Running default===================
------------------- {} -------------------
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Bases_double_ulong() : 16
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Bases_long_ulong() : 16
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Bases_ulong_ulong() : 16
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Indices_float_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Indices_int_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Indices_uint_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Indices_float_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Indices_int_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Indices_uint_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Indices_double_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Indices_long_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Indices_ulong_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Indices_double_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Indices_long_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorFirstFaulting_Indices_ulong_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorByteZeroExtendFirstFaulting_Indices_int_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorByteZeroExtendFirstFaulting_Indices_int_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorByteZeroExtendFirstFaulting_Indices_long_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorByteZeroExtendFirstFaulting_Indices_long_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorByteZeroExtendFirstFaulting_Indices_uint_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorByteZeroExtendFirstFaulting_Indices_uint_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorByteZeroExtendFirstFaulting_Indices_ulong_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorByteZeroExtendFirstFaulting_Indices_ulong_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16SignExtendFirstFaulting_Indices_int_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16SignExtendFirstFaulting_Indices_int_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16SignExtendFirstFaulting_Indices_long_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16SignExtendFirstFaulting_Indices_long_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16SignExtendFirstFaulting_Indices_uint_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16SignExtendFirstFaulting_Indices_uint_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16SignExtendFirstFaulting_Indices_ulong_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16SignExtendFirstFaulting_Indices_ulong_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16WithByteOffsetsSignExtendFirstFaulting_offsets_int_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16WithByteOffsetsSignExtendFirstFaulting_offsets_int_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16WithByteOffsetsSignExtendFirstFaulting_offsets_long_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16WithByteOffsetsSignExtendFirstFaulting_offsets_long_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16WithByteOffsetsSignExtendFirstFaulting_offsets_uint_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16WithByteOffsetsSignExtendFirstFaulting_offsets_uint_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16WithByteOffsetsSignExtendFirstFaulting_offsets_ulong_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16WithByteOffsetsSignExtendFirstFaulting_offsets_ulong_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt32SignExtendFirstFaulting_Indices_long_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt32SignExtendFirstFaulting_Indices_long_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt32SignExtendFirstFaulting_Indices_ulong_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt32SignExtendFirstFaulting_Indices_ulong_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt32WithByteOffsetsSignExtendFirstFaulting_offsets_long_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt32WithByteOffsetsSignExtendFirstFaulting_offsets_long_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt32WithByteOffsetsSignExtendFirstFaulting_offsets_ulong_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt32WithByteOffsetsSignExtendFirstFaulting_offsets_ulong_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorSByteSignExtendFirstFaulting_Indices_int_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorSByteSignExtendFirstFaulting_Indices_int_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorSByteSignExtendFirstFaulting_Indices_long_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorSByteSignExtendFirstFaulting_Indices_long_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorSByteSignExtendFirstFaulting_Indices_uint_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorSByteSignExtendFirstFaulting_Indices_uint_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorSByteSignExtendFirstFaulting_Indices_ulong_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorSByteSignExtendFirstFaulting_Indices_ulong_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtendFirstFaulting_offsets_int_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtendFirstFaulting_offsets_int_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtendFirstFaulting_offsets_long_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtendFirstFaulting_offsets_long_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtendFirstFaulting_offsets_uint_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtendFirstFaulting_offsets_uint_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtendFirstFaulting_offsets_ulong_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16WithByteOffsetsZeroExtendFirstFaulting_offsets_ulong_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Indices_int_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Indices_int_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Indices_long_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Indices_long_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Indices_uint_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Indices_uint_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Indices_ulong_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Indices_ulong_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtendFirstFaulting_offsets_int_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtendFirstFaulting_offsets_int_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtendFirstFaulting_offsets_long_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtendFirstFaulting_offsets_long_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtendFirstFaulting_offsets_uint_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtendFirstFaulting_offsets_uint_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtendFirstFaulting_offsets_ulong_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32WithByteOffsetsZeroExtendFirstFaulting_offsets_ulong_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Indices_int_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Indices_int_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Indices_long_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Indices_long_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Indices_uint_int() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Indices_uint_uint() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Indices_ulong_long() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Indices_ulong_ulong() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorByteZeroExtendFirstFaulting_Bases_long_ulong() : 16
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorByteZeroExtendFirstFaulting_Bases_ulong_ulong() : 16
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16SignExtendFirstFaulting_Bases_long_ulong() : 16
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorInt16SignExtendFirstFaulting_Bases_ulong_ulong() : 16
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Bases_long_ulong() : 16
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt16ZeroExtendFirstFaulting_Bases_ulong_ulong() : 16
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Bases_long_ulong() : 16
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_GatherVectorUInt32ZeroExtendFirstFaulting_Bases_ulong_ulong() : 16
===================Running jitstress===================
------------------- {'JitMinOpts': '1'} -------------------
------------------- {'JitStress': '1'} -------------------
------------------- {'JitStress': '2'} -------------------
------------------- {'JitStress': '1', 'TieredCompilation': '1'} -------------------
------------------- {'JitStress': '2', 'TieredCompilation': '1'} -------------------
------------------- {'TailcallStress': '1'} -------------------
------------------- {'ReadyToRun': '0'} -------------------
===================Running jitstressregs===================
------------------- {'JitStressRegs': '1'} -------------------
------------------- {'JitStressRegs': '2'} -------------------
------------------- {'JitStressRegs': '3'} -------------------
------------------- {'JitStressRegs': '4'} -------------------
------------------- {'JitStressRegs': '8'} -------------------
------------------- {'JitStressRegs': '0x10'} -------------------
------------------- {'JitStressRegs': '0x80'} -------------------
------------------- {'JitStressRegs': '0x1000'} -------------------
------------------- {'JitStressRegs': '0x2000'} -------------------
===================Running jitstress2-jitstressregs===================
------------------- {'JitStress': '2', 'JitStressRegs': '1'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '2'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '3'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '4'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '8'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x10'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x80'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x1000'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x2000'} -------------------
```
</details>

a74nh commented:
```
src\libraries\System.Private.CoreLib\src\System\Runtime\Intrinsics\Arm\Sve.PlatformNotSupported.cs(3703,1): error CS1570: (NETCORE_ENGINEERING_TELEMETRY=Build) XML comment has badly formed XML -- 'Expected an end tag for element 'uint'.'
```
 - looks like valid errors from the CI

a74nh commented on src/libraries/Common/tests/TestUtilities/System/Buffers/BoundedMemory.Windows.cs:
```diff
@@ -295,7 +295,7 @@ internal static VirtualAllocHandle Allocate(IntPtr lpAddress, IntPtr dwSize, Vir
         }
 
-        private static partial class UnsafeNativeMethods
+        private static class UnsafeNativeMethods
```
Intentional change?

SwapnilGaikwad replied:
ooh, nope. Removed it.

SwapnilGaikwad committed:
Fix the formatting of a comment causing build failure

SwapnilGaikwad committed:
Merge main

SwapnilGaikwad committed:
Fix unintentional change

amanasifkhalid commented on src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs:
```diff
@@ -3665,7 +3665,89 @@
+    ("SveGatherVectorFirstFaultingIndices.template",   new Dictionary<string, string> { ["TestName"] = "Sve_GatherVectorInt16SignExtendFirstFaulting_Indices_uint_uint",                                                     ["Isa"] = "Sve",           ["LoadIsa"] = "Sve",     ["Method"] = "GatherVectorInt16SignExtendFirstFaulting",                                          ["RetVectorType"] = "Vector" ...
+    ("SveGatherVectorFirstFaultingIndices.template",   new Dictionary<string, string> { ["TestName"] = "Sve_GatherVectorInt16SignExtendFirstFaulting_Indices_ulong_long",                                                    ["Isa"] = "Sve",           ["LoadIsa"] = "Sve",     ["Method"] = "GatherVectorInt16SignExtendFirstFaulting",                                          ["RetVectorType"] = "Vector" ...
+    ("SveGatherVectorFirstFaultingIndices.template",   new Dictionary<string, string> { ["TestName"] = "Sve_GatherVectorInt16SignExtendFirstFaulting_Indices_ulong_ulong",                                                   ["Isa"] = "Sve",           ["LoadIsa"] = "Sve",     ["Method"] = "GatherVectorInt16SignExtendFirstFaulting",                                          ["RetVectorType"] = "Vector" ...
+    ("SveGatherVectorFirstFaultingIndices.template",   new Dictionary<string, string> { ["TestName"] = "Sve_GatherVectorInt16WithByteOffsetsSignExtendFirstFaulting_offsets_int_int",                                        ["Isa"] = "Sve",           ["LoadIsa"] = "Sve",     ["Method"] = "GatherVectorInt16WithByteOffsetsSignExtendFirstFaulting",                           ["RetVectorType"] = "Vector" ...
```
For the `WithByteOffsets` APIs, should we be using `SveGatherVectorFirstFaultingIndices.template`, or do we need a new template that works specifically with byte offsets?

SwapnilGaikwad replied:
You're right that the names maybe a little misleading. The `*FirstFaultingIndices.template` take a pointer in op2 and indices in op3 while `*FirstFaultingBases.template` take a vector of addresses.
Would you suggest to change their names to something different?

amanasifkhalid replied:
I think the naming is fine. My concern is this template is treating `op3` like a Vector of indices, instead of a Vector of byte offsets. For example, suppose the data source is an array of `Int64`. In this template's implementation of `RunBasicScenario_LoadFirstFaulting`, we set the last element of `op3` to `outsideBoundsIndex` to force a fault. `outsideBoundsIndex` is equal to `(Vector<{Int64}>.Count - 1)`, so with a 128-bit width, it's `outsideBoundsIndex = 2 - 1 = 1`. An index of 1 would be out-of-bounds, but a byte offset of 1 shouldn't be, so I'd expect this to not fault. To cause a fault, I think we'd need to set the last element of `op3` to `outsideBoundsIndex * sizeof(Int64)`, or 8 bytes.

I think you can reuse the new template I'm adding in #106199 to test the byte offset APIs.

SwapnilGaikwad replied:
Agreed, makes sense. I'll let your patch merge and then update this PR.

amanasifkhalid commented:
@SwapnilGaikwad #106199 has been merged

SwapnilGaikwad committed:
Merge main

Change-Id: Ia064b10db27b63b1b1254c4f478bf11a81aff4b2

SwapnilGaikwad committed:
Use a template for byte offset

a74nh commented on src/coreclr/jit/hwintrinsiclistarm64sve.h:
```diff
@@ -111,16 +111,26 @@ HARDWARE_INTRINSIC(Sve,           GatherPrefetch64Bit,
 HARDWARE_INTRINSIC(Sve,           GatherPrefetch8Bit,                                               -1,     -1,     {INS_sve_prfb,       INS_sve_prfb,       INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid},     HW_Category_Special,               HW_Flag_Scalable|HW_Flag_BaseTypeFromFirstArg|HW_ ...
 HARDWARE_INTRINSIC(Sve,           GatherVector,                                                     -1,     -1,     {INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_sve_ld1w,       INS_sve_ld1w,       INS_sve_ld1d,       INS_sve_ld1d,       INS_sve_ld1w,       INS_sve_ld1d},    HW_Category_MemoryLoad,            HW_Flag_Scalable|HW_Flag_SpecialCodeGen|HW_Flag_E ...
 HARDWARE_INTRINSIC(Sve,           GatherVectorByteZeroExtend,                                       -1,     -1,     {INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_sve_ld1b,       INS_sve_ld1b,       INS_sve_ld1b,       INS_sve_ld1b,       INS_invalid,        INS_invalid},     HW_Category_MemoryLoad,            HW_Flag_Scalable|HW_Flag_SpecialCodeGen|HW_Flag_E ...
+HARDWARE_INTRINSIC(Sve,           GatherVectorByteZeroExtendFirstFaulting,                          -1,     -1,     {INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_sve_ldff1b,     INS_sve_ldff1b,     INS_sve_ldff1b,     INS_sve_ldff1b,     INS_invalid,        INS_invalid},     HW_Category_MemoryLoad,            HW_Flag_Scalable|HW_Flag_SpecialCodeGen|HW_Flag_E ...
```
Could you please rebase to main (again!) and add `HW_Flag_ZeroingMaskedOperation` to all these new APIs. That flag needs adding to any API where the predicate is always `Pg/Z`.

SwapnilGaikwad replied:
Done!

SwapnilGaikwad committed:
Merge main

JulieLeeMSFT commented:
@amanasifkhalid, @dotnet/arm64-contrib, please do code review for this last FFR PR.

amanasifkhalid commented on src/coreclr/jit/hwintrinsiclistarm64sve.h:
```diff
@@ -111,16 +111,26 @@ HARDWARE_INTRINSIC(Sve,           GatherPrefetch64Bit,
 HARDWARE_INTRINSIC(Sve,           GatherPrefetch8Bit,                                               -1,     -1,     {INS_sve_prfb,       INS_sve_prfb,       INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid},     HW_Category_Special,               HW_Flag_Scalable|HW_Flag_BaseTypeFromFirstArg|HW_ ...
 HARDWARE_INTRINSIC(Sve,           GatherVector,                                                     -1,     -1,     {INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_sve_ld1w,       INS_sve_ld1w,       INS_sve_ld1d,       INS_sve_ld1d,       INS_sve_ld1w,       INS_sve_ld1d},    HW_Category_MemoryLoad,            HW_Flag_Scalable|HW_Flag_SpecialCodeGen|HW_Flag_E ...
 HARDWARE_INTRINSIC(Sve,           GatherVectorByteZeroExtend,                                       -1,     -1,     {INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_sve_ld1b,       INS_sve_ld1b,       INS_sve_ld1b,       INS_sve_ld1b,       INS_invalid,        INS_invalid},     HW_Category_MemoryLoad,            HW_Flag_Scalable|HW_Flag_SpecialCodeGen|HW_Flag_E ...
+HARDWARE_INTRINSIC(Sve,           GatherVectorByteZeroExtendFirstFaulting,                          -1,     -1,     {INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_sve_ldff1b,     INS_sve_ldff1b,     INS_sve_ldff1b,     INS_sve_ldff1b,     INS_invalid,        INS_invalid},     HW_Category_MemoryLoad,            HW_Flag_Scalable|HW_Flag_SpecialCodeGen|HW_Flag_E ...
```
Should these use `HW_Flag_SpecialSideEffectMask`, or should they be switched over to `HW_Flag_SpecialSideEffect_Other`. I think #105973 moved all the first-faulting intrinsics over to the latter flag.

SwapnilGaikwad replied:
You're right Aman, I totally missed the updates from https://github.com/dotnet/runtime/pull/105973. Updated them now.

amanasifkhalid reviewed (commented):
Sorry to ask you to do this, but could you try uncommenting the `ConditionalSelect_FalseOp` scenarios like I did in #106349, and re-run the tests you added, please? Just want to make sure that PR can go in cleanly alongside this one.

SwapnilGaikwad committed:
Update HW_Flag_SpecialSideEffectMask to HW_Flag_SpecialSideEffect_Other for firstFaulting APIs

SwapnilGaikwad committed:
Merge main

amanasifkhalid reviewed (approved):
LGTM, verified offline that `ConditionalSelect_FalseOp` works too. Thanks!
